### PR TITLE
BREAKING: Rework card config system (supports O3-3242)

### PR DESCRIPTION
**Pull request by brandones**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

Simplifies the config schema. No more polymorphic config objects. Makes each card row other than the header and footer into an extension. Right now there's just the "coloredObsTags" extension, which is used for displaying risk factors.

This means that card rows should be configured using the extension configuration system. This is the one thing that is a little less intuitive than it was before, since the row config lives somewhere separate from the card config. However, what we get from the trade-off is that new types of rows can be introduced by third-party modules without needing to modify esm-ward-app.

The new KGH config should be

```
{
  "@openmrs/esm-ward-app": {
    "wardPatientCards": {
      "obsElementDefinitions": [
        {
          "conceptUuid": "3cee82de-26fe-102b-80cb-0017a47871b2",
          "id": "gravida"
        },
        {
          "id": "admission-reason",
          "conceptUuid": "86a2cf11-1ea5-4b8a-9e4b-08f4cdbe1346",
          "onlyWithinCurrentVisit": true,
          "orderBy": "ascending",
          "label": "REASON"
        }
      ],
      "addressElementDefinitions": [
        {
          "id": "patient-address",
          "fields": [
            "countyDistrict"
          ]
        }
      ],
      "cardDefinitions": [
        {
          "id": "default",
          "headerRowElements": [
            "patient-age",
            "patient-address",
            "patient-identifier",
            "admission-reason"
          ],
          "footerRowElements": [
            "gravida"
          ],
          "patientAddressElementFields": [
            "countyDistrict"
          ],
          "appliedTo": []
        }
      ]
    }
  },
  "colored-obs-tags-card-row": {
    "conceptUuid": "447f14be-4f34-4cba-9f80-8e091e23c6e3",
    "summaryLabel": "Pregancy Complication(s)",
    "tags": [
      {
        "color": "red",
        "appliedToConceptSets": [
          "447f14be-4f34-4cba-9f80-8e091e23c6e3"
        ]
      }
    ]
  }
}
```

No UI change.

## Screenshots

image

## Related Issue

https://issues.openmrs.org/browse/O3-3242

## Other

I'm not 100% sure the admission-reason config is working & correct because none of the admitted patients seem to have the relevant obs.

It wouldn't be unreasonable, I think, to rename `colored-obs-tags-card-row` to `colored-obs-tags`, since there's actually nothing very specific to the card row situation about it. It could just as easily be dropped in to lots of places. The card itself is providing the row styling.

NOTE: I've left all the file names intact in order to facilitate reviewing. Once this approach is something we're happy moving forward with, I'll rename files and folders into a structure more appropriate for the new setup.

### CODE REVIEW

File: packages/esm-ward-app/src/config-schema.ts
Comment (chibongho) [inline packages/esm-ward-app/src/config-schema.ts:153]: I think `patientAddressElementFields` should belong here instead of the `WardPatientCardDefinition` interface.

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx
Comment (chibongho) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx:61]: It feels off to have `WardPatientIdentifier` appear twice in this switch statement, one for having it take the default config and one for having it take a custom config.

I think the `id` field of `obsElementDefinition` / `identifierElementDefinition` is supposed to specify which element it configures. Can we have the `id` be nullable, in which case it applies the config to the default element (patient-identifier / patient-obs)? I think this should simplify the logic here.

**Review by chibongho — COMMENTED** [review]
This simplifies the config a bit. Some things here don't feel right, but I think they are fixable.

Previously, I envisioned the config system to allow for arbitrary number of rows, each row can have any of the predefined elements, or even extension elements (not yet implemented). A custom "row" can technically be implemented as a row with an extension element. If I understand correctly, the changes in this PR are:
- we treat each element type defined here as first-class objects.
- The header and footer rows get first class support. All other rows sandwiched in between are to be added as extensions.
- The `WardPatientCodedObsTags` is so specific and unique that it's elevated from being an element to being an extension row, with its own extension config. (This feels strange, but I don't feel too strongly about it)

Some questions I have:
- How would someone add an extension element to the header and footer? (somewhat likely) We need to consider what props we should pass into it, and how we define the extension slot name.
- How would someone add predefined elements into the sandwiched extension rows? (perhaps less likely) Would it only be possible if we also make the predefined elements available as extensions?

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx
Comment (ibacher) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx:40]: Minor, but something which renders a React element, as opposed to data-fetching, feels like it should be in a `.component.tsx` file rather than a `.resource.tsx` file (even though this MFE doesn't do a great job of following the [Coding Conventions](https://openmrs.atlassian.net/wiki/spaces/projects/pages/151061068/Coding+Conventions)).

**Review by ibacher — COMMENTED** [review]
I don't feel I'm well-placed to evaluate this commit. However, broadly speaking, I'm in favour of anything that keeps the configuration relatively simple as a design goal of O3 is to have it configurable by implementers with minimal programming knowledge.

**Comment by mogoodrich** [comment]
I'll generally defer to @chibongho  has he is more up-to-speed on the original design of this, but I'll keep an eye on the PR and weigh in on things I see.

A few thoughts:
* I do agree that being able to insert rows via extension is a good idea (was this not possible before)
* But I do thing we need to able to apply custom elements to the header/footer (and that this may be a common use case, as @chibongho notes); though, fwiw, I suppose it would be enough to support replacing the entire header row with a custom row (which, practically, would just be a matter of having a property to suppress the default header or footer)?

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx
Comment (brandones) [reply packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx:40]: Yeah certainly, lots of files will get renamed after the dust settles a little, before I merge this PR.

File: packages/esm-ward-app/src/config-schema.ts
Comment (brandones) [reply packages/esm-ward-app/src/config-schema.ts:153]: Sounds good. That does mean that the same address elements will necessarily be used for every ward, but that's probably fine.

Comment (chibongho) [reply packages/esm-ward-app/src/config-schema.ts:152]: Wait... why? Each value in `obsElementDefinitions` and `identifierElementDefinitions` can specify which element id to apply the config to. Can't we just do the same thing with `patientAddressElementFields`?

**Comment by brandones** [comment]
Thanks so much for your thoughtful reviews, @chibongho , @mogoodrich , and @ibacher .

> The `WardPatientCodedObsTags` is so specific and unique that it's elevated from being an element to being an extension row, with its own extension config. (This feels strange, but again I don't feel too strongly about it)

Yeah, the idea is that each row is an extension—so the baby row, when it's time to add that, will also be an extension.

> How would someone add an extension element to the header and footer? (somewhat likely) We need to consider what props we should pass into it, and how we define the extension slot name.

Good thought! I've added extension slots for those.

> How would someone add predefined elements into the sandwiched extension rows? (perhaps less likely) Would it only be possible if we also make the predefined elements available as extensions?

Right now the only rows that are in any designs I've seen are quite different than the contents of the header and footer. The expectation encoded in this approach is that rows will be coded from scratch. I think this is the most appropriate way about it, because all of the existing row designs are highly idiosyncratic, and each row is pretty small and self-contained, easy to build from scratch. If someone really needs a row that displays the value of an obs in plain text, and for some reason they don't want to just put it in the header, they'll have to write a fetcher for that obs, and that's probably okay. It won't pollute any other code anyway.

File: packages/esm-ward-app/src/config-schema.ts
Comment (brandones) [reply packages/esm-ward-app/src/config-schema.ts:152]: Oh I see, is what you're suggesting that instead of having 'patient-address' be a built-in element that is configurable, that instead we should provide a system for defining elements of type 'patient-address' which can then be included, but where the only real difference between the different patient-address elements is which fields are shown?

Comment (chibongho) [reply packages/esm-ward-app/src/config-schema.ts:152]: Yeah. That follows the same pattern as the `identiferElement` and `obsElement`, right?

File: packages/esm-ward-app/src/config-schema-extension-colored-obs-tags.ts
Comment (brandones) [reply packages/esm-ward-app/src/config-schema-extension-colored-obs-tags.ts:1]: Yes, other row extensions will have their own extension schema. Whether they all go in the same file or not, we don't really have a precedent. I could go either way.

Comment (chibongho) [reply packages/esm-ward-app/src/config-schema-extension-colored-obs-tags.ts:1]: Sorry... ignore this. I wrote this when I didn't realize you were turning coloredObsTagElement into its own row / extension.

File: packages/esm-ward-app/src/config-schema.ts
Comment (brandones) [reply packages/esm-ward-app/src/config-schema.ts:152]: Sure, it does.

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx
Comment (brandones) [reply packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx:61]: No, the config system doesn't support types being nullable. `null` can be provided as a default value, but an element with type `Type.String` cannot have `null` provided as the value by configuration.

I'm not sure what feels off about this, but it seems like a more intuitive API to provide than one in which there is a magic value of `id` that modifies the default identifier element.

Comment (brandones) [reply packages/esm-ward-app/src/ward-patient-card/ward-patient-card-row.resources.tsx:61]: Oh I think I figured out a slightly better way about it

**Comment by brandones** [comment]
Ok, I've updated the address and identifier configurations. I'll fix the merge conflicts shortly.

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx
Comment (chibongho) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx:64]: I was hoping there is an `ExtensionElement` that can be placed with other elements in arbitrary instead of at the end, but we can do that in a different PR.

**Review by chibongho — APPROVED** [review]
Looks ok. Pretty scary change TBH. Hoping that nothing breaks (or that they are easy fixes).

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx
Comment (brandones) [reply packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx:64]: I don't understand, could you elaborate on this?

**Comment by brandones** [comment]
I've updated the config in the PR description; that should work. Thanks @chibongho !

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx
Comment (chibongho) [reply packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx:64]: Sorry, typo. should be "in arbitrary order". So we can have, for example, `bed-number`, `patient-name`, `extension-element`, `patient-identifier` (instead of having the `extension-element` at the end)

Comment (brandones) [reply packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx:64]: If that becomes a real need, the thing to do would probably be to turn the elements that come after the patient name into extensions and have an extension slot come after name. I would suggest not treating bed number and patient name as arbitrary "bento box" elements, but rather as things that are managed by the card itself. So in the main card you'd have

bed number    |   patient name    |  header extension slot

And then in the admission request card you'd have

patient name    |  header extension slot

And then all of those elements, obs display etc, would be configurable directly using extension config (so the card wouldn't have anything to do with their configuration), and you'd just use the config system to add/remove/rearrange them.

But that would come with some performance cost (which would be significantly lightened by [O3-1418](https://openmrs.atlassian.net/browse/O3-1418)). So I'd recommend not doing that until it's clear that there is a use case.

**Comment by chibongho** [comment]
> I've updated the config in the PR description; that should work. Thanks @chibongho !

@brandones the new config seems to work. thanks!

**Pull request merged**
